runtime/compiler/env/VMJ9.cpp

```diff
@@ -4858,6 +4859,37 @@ TR_J9VMBase::targetMethodFromMethodHandle(uintptr_t methodHandle)
    return targetMethodFromMemberName(vmentry);
    }
 
+TR::KnownObjectTable::Index
+TR_J9VMBase::getKnotIndexOfInvokeCacheArrayAppendixIndex(TR::Compilation *comp, uintptr_t *invokeCacheArray)
+   {
+   TR::KnownObjectTable *knot = comp->getOrCreateKnownObjectTable();
+   if (!knot) return TR::KnownObjectTable::UNKNOWN;
+
+   TR::VMAccessCriticalSection vmAccess(this);
+   uintptr_t appendixElementRef = (uintptr_t) getReferenceElement(*invokeCacheArray, JSR292_invokeCacheArrayAppendixIndex);
+   return knot->getOrCreateIndex(appendixElementRef);
+   }
+
+TR_ResolvedMethod *
+TR_J9VMBase::targetResolvedMethodFromInvokeCacheArray(TR::Compilation *comp, TR_ResolvedMethod *owningMethod, uintptr_t *invokeCacheArray)
+   {
+   TR_OpaqueMethodBlock *targetMethodObj = 0;
+      {
+      TR::VMAccessCriticalSection vmAccess(this);
+      uintptr_t memberNameElementRef = (uintptr_t) getReferenceElement(*invokeCacheArray, JSR292_invokeCacheArrayMemberNameIndex);
+      targetMethodObj = targetMethodFromMemberName(memberNameElementRef);
+      }
+   return createResolvedMethod(comp->trMemory(), targetMethodObj, owningMethod);
+   }
+
+TR::SymbolReference*
+TR_J9VMBase::refineInvokeCacheElementSymRefWithKnownObjectIndex(TR::Compilation *comp, TR::SymbolReference *originalSymRef, uintptr_t *invokeCacheArray, int32_t arrayIndex)
+   {
+   TR::VMAccessCriticalSection vmAccess(this);
+   uintptr_t arrayElementRef = (uintptr_t) getReferenceElement(*invokeCacheArray, arrayIndex);
+   return comp->getSymRefTab()->refineInvokeCacheElementSymRefWithKnownObjectIndex(originalSymRef, arrayElementRef);
```

Comment:
If the refinement of the symref has to be done via a call to a J9VM helper to make this work in JITServer, we should just move the implementation of `J9::SymbolReferenceTable::refineInvokeCacheElementSymRefWithKnownObjectIndex` here.

Reply:
It's not that we can't do it from `J9::SymbolReferenceTable::refineInvokeCacheElementSymRefWithKnownObjectIndex`, we could, but I want to avoid placing remote calls outside of front-end queries as much as possible.
If you are fine with moving the implementation here, I'll do it.